### PR TITLE
New Javascript code-generation

### PR DESCRIPTION
**Pull request by alpmestan**
This PR contains the work done in #159 to supply javascript code generators based on vanilla Javascript and Angular in addition to jQuery, and cleans up a lot of the code to provide a dead simple API that boils down to using `jsForAPI`/`writeJSForAPI` along with the generator you want. That's it.

We also thought this would be a good opportunity to rename the package from `servant-jquery` to `servant-js`, and same with the module names.

I might have missed a few `s/servant-jquery/servant-js/` even though I think there's none of that left anywhere important.

PS: do not merge yet, I yet have to write the chanlog entry for this.

### CODE REVIEW

File: servant-js/test/Servant/JSSpec.hs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        parseFromString jsText `shouldSatisfy` isRight
+        jsText `shouldContain` "headerXWhatsForDinner"
+        jsText `shouldContain`  (header n "X-WhatsForDinner" $ "\"I would like \" + headerXWhatsForDinner + \" with a cherry on top.\"")
+
+    it "can generate the whole javascript code string at once with jsForAPI" $ do
+        let jsStr = jsForAPI (Proxy :: Proxy TestAPI) (concat . map gen)
+        parseFromString jsStr `shouldSatisfy` isRight
+    where
+        specLabel = "generateJS(" ++ (show n) ++ ")"
+        --output = print
```
Comment (jkarni) [inline servant-js/test/Servant/JSSpec.hs]: Can we remove the commented-out lines?

File: servant-js/test/Servant/JSSpec.hs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        output _ = return ()
+        genJS req = gen req
+        header :: TestNames -> String -> String -> String
+        header v headerName headerValue
+            | v `elem` [Vanilla, VanillaCustom] = "xhr.setRequestHeader(\"" ++ headerName ++ "\", " ++ headerValue ++ ");\n"
+            | otherwise = "headers: { \"" ++ headerName ++ "\": " ++ headerValue ++ " }\n"
+        --header _ _ _ = "Not Implemented"
```
Comment (jkarni) [inline servant-js/test/Servant/JSSpec.hs]: And this one.

File: servant-js/test/Servant/JSSpec.hs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+customOptions :: CommonGeneratorOptions
+customOptions = defCommonGeneratorOptions {
+            successCallback = "okCallback",
+            errorCallback = "errorCallback"
+        }
+                 
+spec :: Spec
+spec = describe "Servant.JQuery" $ do
+    (generateJSSpec Vanilla JS.generateVanillaJS)
```
Comment (jkarni) [inline servant-js/test/Servant/JSSpec.hs]: Any reason for the extra parens?

File: servant-js/src/Servant/JS/Angular.hs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+-- | Generate regular javascript functions that use the $http service.
+angularWith :: AngularOptions -> CommonGeneratorOptions -> JavaScriptGenerator
+angularWith ngopts opts = intercalate "\n\n" . map (generateAngularJSWith ngopts opts)
+
+-- | js codegen using $http service from Angular using default options
+generateAngularJS :: AngularOptions -> AjaxReq -> String
+generateAngularJS ngOpts = generateAngularJSWith ngOpts defCommonGeneratorOptions
+    
+-- | js codegen using $http service from Angular
```
Comment (jkarni) [inline servant-js/src/Servant/JS/Angular.hs]: Doesn't `$` mean a variable in haddock? (Just wondering if anyone took a look at the generated haddocks.)

Comment (alpmestan) [inline servant-js/src/Servant/JS/Angular.hs]: It prints `$http` just like any other "word" just fine here.

**Comment by alpmestan** [comment]
Just pushed a commit that addresses your comments.

File: servant-js/src/Servant/JS/Angular.hs (hunk 4 above)
Comment (freezeboy) [inline servant-js/src/Servant/JS/Angular.hs]: in angular, they call their "modules" $http, $service, etc...
(js guys like the $ sign)

**Comment by jkarni** [comment]
LGTM!

**Comment by alpmestan** [comment]
Hooray, thanks a lot for your work on that @freezeboy, can't wait for your next patches ;)

**Pull request merged**